### DISCUSSION TITLE
Limpieza de datos

### DISCUSSION THREAD

**Discussion by ypne** [IIC2413/Syllabus-2024-2#156] [General]
Este espacio se abre para que los grupos puedan hacer sus consultas sobre el tratamiento de los datos que violen las restricciones de integridad o formatos definidos (y solo eso). No los ingresen en Issues.

**Comment by ypne** [comment]
1.-Construya una aplicaci´on cargadora de datos haciendo uso de PHP. Al igual que la entrega E0 su
grupo deber´a cargar los datos desde los archivos entregados mediante el cargador mencionado, pero
esta vez el destino son las tablas de su base de datos.
2.- El cargador puede ser ejecutado por l´ınea de comando (CLI), no es necesario que sea una aplicaci´on web.
3.-El cargador deber´a: leer todos los datos o una parte de ellos, e insertarlos correctamente en las tablas, validando todas las restricciones de integridad y formatos indicados en el enunciado y reflejadas en su esquema.
4.- Las tuplas que violen las
restricciones de integridad y atributos que violen el formato deber´an ser almacenadas en un archivo
para posterior an´alisis y, si procede, su correcci´on.
Nota: Puede separar el archivo de datos en varios CSV si lo encuentra necesario.
4. Una vez cargados todos los datos que no violen las restricciones, su grupo deber´a analizar los datos
rechazados e intentar recuperar lo m´aximo posible. Las alternativas son:
4.1 • Corregir: si es que ud. tiene informaci´on adicional o alguna buena raz´on puede reemplazar los
datos conflictivos con datos v´alidos (por ejemplo si hay un tel´efono de 8 cifras, se puede anteponer
un 9 o si un correo electr´onico tiene un car´acter inv´alido como el blanco, eliminarlo).
4.2 • Reemplazar: sin el dato no se puede corregir, se puede reemplazar por un valor que sea permitido
por el dominio del atributo pero que simbolice que el valor es nulo y as´ı no se pierden los otros
datos que vienen en la tupla (por ejemplo si el tel´efono contiene una letra, claramente es inv´alido
por lo que se puede reemplazar por 0 o nulo)
4.3 • Descartar: Si no hay otra opci´on, la tupla se debe descartar en cascada.
Nota 1: En algunos atributos ser´a necesario estandarizar los valores a un conjunto de valores
acotados, por ejemplo el ´ultimo logro debe soportar los valores indicados (2 a˜no equivale a nivel 4).
Nota 2: A diferencia de la E0, en esta ocasi´on los datos originales NO SE PUEDEN MODIFICAR, solo se permite modificar el archivo de los datos que violen las restricciones de integridad y los formatos, los cuales deben ser cargados siempre con la aplicaci´on PHP
cargadora de datos.
Nota 3: Es recomendable que su aplicaci´on pueda detectar y corregir
formatos (sanitizaci´on de input)

**Comment by ypne** [comment]
PREREQUISITOS, EQUIVALENCIAS y MALLAS

1.- Cursos: Los cursos tienen 2 componentes los 3 caracteres del programa y un número todos los cursos con el mismo número son equivalentes en los diferentes programas
2.- Prerequisitos: En el tab de prerequisitos está el curso y su prerequisito (solo el número porque todos los cursos con igual número son equivalentes)
3.- El tab malla es solo referencial para que vean el avance y los ciclos, este no es necesario en la base ya que la información está presente en las tablas de asignaturas, etc.

**Comment by ypne** [comment]
Valores nulos
Hay varios atributos en la base que no deben ser nulos, en el modelo ustedes deben exigir que eso ocurra; sin embargo, en los datos vienen muchos tuplas con valores nulos en atributos que no debieran. Para ello debes forzar datos en los valores. Tomemos el caso del RUN: Si  lo usan como llave, no puede ser nulo por lo que el atributo debe ser NOT NULL, pero en los datos pueden venir datos nulos; SOLUCION? buscar el dato en otro archivo (hay datos en varios archivos) o usar valores claramente no válidos (run del rango 60 millones)

**Comment by ypne** [comment]
Pregunta: Hola, una pregunta, ¿puedo limpiar los archivos al mismo tiempo que creo las tablas, por ejemplo con los espacios nulos?¿ O tengo que crear la tabla de errores con absolutamente todos los errores y luego ir como "sumándolos" a la tabla sin errores ?

respuesta:
Primero que nada una definición: Las tablas se crean directo en la base usando psql, no es necesario hacerlo con un programa.
Por otra parte,  entendiendo que por crear se refieren a cargar, si se puede hacer para los temas de formato, o sea caracteres iextraños en textos (letras en números telefónicos, blancos en correos o el formato del correo electrónico. Lo que no se puede hacer (porque es tarea de la base de datos) son las restricciones de integridad de los datos como las FK que referencian valores que no existen, valores nulos en campos obligatorios, etc.

**Reply by emivenezian** [reply]
mm, mi pregunta es por ejemplo para la limpieza el orden que hay que seguir. Por ejemplo, para la E0 tenia funciones que verificaban q el rut cumpliera con ser int, 8 digitos, sin espacios, sin repetidos, etc, y retornaba un array solo con los datos limpios y filtrados. 
1. Por lo que entiendo del enunciado, esa funcion ya no sirve para esta entrega. Deberia tener una funcion que me "retorne" unicamente los ruts que cumplen con lo pedido en el enunciado, y todo lo demas, guardarlo en un archivo. Luego en este archivo realizar las correcciones. 
2. Esto no me hace sentido, pues es hacer practicamente lo mismo dos veces, es muy poco eficiente. No entiendo porque no podemos simplemente corregir los datos y cargar aquellos que si cumplen con las condiciones en las tablas, me es mucho mas logico. 
3. Y tampoco me queda claro lo de no poder modificar los archivos, leerlos y guardar su info en arrays no es modificarlos cierto? como se hizo en la E0.

**Reply by ypne** [reply]
Todo lo que sea validación del input se puede reutilizar, es decir, que cada atributo tenga el FORMATO sea correcto. Por ejemplo que el RUN sean solo número, que el DV sea un caracter. OJO el RUN no tiene 8 dígitos, tiene MÁXIMO 8.

La validación de las restricciones de integridad las hace directamente el DBMS al hacer el INSERT en la base de datos, por ejemplo hay duplicados además valida las FK, TIPO de DATO, NOT NULL, UNIQUE, etc.

No hay nada que repetir
No se usan arreglos
No se pueden corregir los datos en el archivo de entrada
lean las condiciones en el enunciado.
"Nota 1: En algunos atributos ser´a necesario estandarizar los valores a un conjunto de valores
acotados, por ejemplo el ´ultimo logro debe soportar los valores indicados (2 a˜no equivale a nivel 4).
Nota 2: A diferencia de la E0, en esta ocasi´on los datos originales NO SE PUEDEN MODIFICAR, solo se permite modificar el archivo de los datos que violen las restricciones de integridad y los formatos, los cuales deben ser cargados siempre con la aplicaci´on PHP
cargadora de datos.
Nota 3: Es recomendable que su aplicaci´on pueda detectar y corregir
formatos (sanitizaci´on de input)"

**Reply by emivenezian** [reply]
entiendo lo que dices, pero aun no me queda claro que se nos pide entonces. En el enunciado sale claramente que "Las tuplas que violen las restricciones de integridad y atributos que violen el formato deber ́an ser almacenadas en un archivo para posterior an ́alisis y, si procede, su correcci ́on". A esto me refiero con hacerlo dos veces, primero tenemos que cargar en las tablas los valores que si cumplen el formato, y luego corregir los que tenian irregularidades y cargarlos en las tablas. Eso es hacerlo dos veces. Es una contradiccion pedir que primero separemos los archivos en validos y no validos (hay q filtarlos) y luego revisamos los invalidos (hay q filtrarlos nuevamente) en un solo paso, no calza.

**Reply by ypne** [reply]
1.- mientras no traten de insertar la tupla (INSERT) no van a saber si la tupla cumple con la restricciones de integridad.
2.- Si cumple la tupla queda insertada, sinó hay que guardarla para post revisión
3.- Una vez modificadas todas las tuplas con problemas se vuelven a cargar(solo esaas) con el mismo cargador
Los archivos NO SE DEBEN separar este proceso lo deben hacer en el cargados usando las funciones del DBMS

**Reply by oliviallanosj** [reply]
Respecto a esto, podemos hacer la limpieza de todos los datos (creando nuevos archivos que son los modificados) y al final, cunado estén todos limpios y también los incorrectos arreglados, los transformamos a tuplas para poder insertarlos a las tablas? El sentido es el mismo. Muchas gracias.

**Comment by ignanuyens** [comment]
Hola, yo tengo una duda en la siguiente parte:
4. Una vez cargados todos los datos que no violen las restricciones, su grupo deber´a analizar los datos
rechazados e intentar recuperar lo m´aximo posible. Las alternativas son:
4.1 [...]
4.2 [...]
4.3 • Descartar: Si no hay otra opci´on, la tupla se debe descartar en cascada.
Nota 1: En algunos atributos ser´a necesario estandarizar los valores a un conjunto de valores
acotados, por ejemplo el ´ultimo logro debe soportar los valores indicados (2 a˜no equivale a nivel 4).
Nota 2: A diferencia de la E0, en esta ocasi´on los datos originales NO SE PUEDEN MODIFICAR, solo se permite modificar el archivo de los datos que violen las restricciones de integridad y los formatos, los cuales deben ser cargados siempre con la aplicaci´on PHP
cargadora de datos.
Nota 3: Es recomendable que su aplicaci´on pueda detectar y corregir
formatos (sanitizaci´on de input)

Mi duda es la siguiente: yo debo intentar cargar mis tablas, para eso yo tengo múltiples "restricciones" para cada atributo. Pongamos el ejemplo que quiero armar una tabla con la hoja de Estudiantes. Iniciaré recorriendo cada tupla, donde para cada una de ellas quiero verificar que los atributos cumplan con mis restricciones. PREGUNA 1: puedo verificar esas restricciones con funciones que cree a parte? es decir, que yo verifique si cumple las condiciones (que son a parte de las restricciones de integridad que existirán al intentar cargar la tupla en la tabla), y en caso que no lo cumpla, la añado
a un archivo que tendrá las tuplas con errores, PREGUNA 2: puedo modificar la tupla ahí mismo, para posteriormente cargarla a la tabla?
La necesitad de crear estas funciones que restringen, a parte de las restricciones de integridad de las tablas, las considero necesarias para sobre todo estandarizar los valores y el formato que tendrán los atributos en las tablas (Nota 1).
PREGUNA 3: no entendí la nota 3, por lo que agradecería su explicación. Ya que escuché que las validaciones se pueden realizar también con SQL y no entiendo cómo sería eso.

Muchas gracias de antemano

**Reply by ypne** [reply]
Pregunta 1: puedo verificar esas restricciones con funciones que cree a parte? 
Respuesta 1: NO ES NECESARIO, esa función la hace automáticamente el DBMS al hacer INSERT. Diferente es el tema de los campos  de texto(nombres,  email) , ya que esos sí se pueden limpiar en PHP(ver Nota3). 
Pregunta 2:  puedo modificar la tupla ahí mismo, para posteriormente cargarla a la tabla?
Respuesta 2: si te refieres a corregir los errores de formato, lo puedes hacer con PHP al leer la tupla. Por ejemplo es psuedo código.
While not EOF (archivo XX.csv)
           Read line (atributo1,..., atributoN)
          for i in 1..N do 
                  Chequeo y corrijo el formato del atributo i
          IF  EXECUTESQL (INSERT en Tabla T (atributo1,..,atributoN)) = ERROR THEN escriba error en errores.txt (Atributo1,.., Atributo N)
Pregunta 3: no entendí la nota 3,
Respuesta: Existen 3 casos posibles:
1.- La tupla está correcta para insertar, en este caso se inserta
2.- La tupla tiene problemas de formato (por ejemplo email con blancos, caracteres extraños, números en los nombres,etc.) En este caso, si se puede arreglar, se arregla. Si no se puede se asigna algún valor que represente nulo y en último caso si lo anterior no se puede la tupla no se carga
3.- La tupla viola restricciones de integridad (por ejemplo largo del telefono diferente de 9 cifras, FK que no existe, etc.) en ese caso se siguen las mismas reglas anteriores.

**Reply by ignanuyens** [reply]
Sigo son dudas. 
Respecto a Pregunta 1. Lo que quiero hacer es leer el archivo original. Ir leyendo las tuplas y comprobar si cumplen con ciertas restricciones (tanto de valor y formato) a través de funciones que yo cree, y en caso que no lo haga, lo agrego a un archivo de errores de dicha hoja (ej: errores_estudiantes). Luego, intento solucionar el error (por ejemplo, si es de formato lo transformo al formato que quiero), y lo agrego a un nuevo archivo que tiene las tuplas arregladas (ej: arreglados_estudiante). Así para cada hoja, por lo que al final tendré 3 hojas para cada una (la original, la con errores y la arreglada). Una vez realizado todo lo anterior, poblaré mis tablas con las tuplas arregladas del archivo arreglados_hoja. 

Realizar esto me permite cargar las tablas de forma más lógica. También, cumpliría con los requisitos del enunciado, por lo que estaría bien, pero me confunden las issues. Quiero que se entienda que me refiero a hacer un filtrado ANTES del filtrado de las restricciones de integridad de las tablas.

muchas gracias de antemano

**Reply by ypne** [reply]
Pregunta 1. Lo que quiero hacer es leer el archivo original. Ir leyendo las tuplas y comprobar si cumplen con ciertas restricciones (tanto de valor y formato) a través de funciones que yo cree.
Respuesta: Esto ya se ha respondido antes, hay varias formas válidas para hacer esto. Los errores de formato se pueden corregir en PHP y cargar a la tabla. Las restricciones de integridad como Violaciones a UNIQUE, FK, NOT NULL etc. las valida automáticamente el DBMS al hacer el INSERT.

**Comment by ypne** [comment]
Pregunta: Hola, podemos asumir que los datos repetidos se eliminan con sql una vez que se suben a las tablas? o en el limpiador de datos tenemos que eliminar los repetidos?

Respuesta: Asumo a que te refieres a tuplas completas repetidas. En este caso en general hay una llave y por definición las llaves tienen la restricción UNIQUE, o sea no puede haber un valor de llave repetido.
la validación de las RESTRICCIONES DE INTEGRIDAD las hace el DBMS en forma automática, no es necesario hacer una limpieza de esos datos en PHP. Es decir si llave(tupla t1)=llave(tuplat2) e INSERT(t1) va antes de INSERT(t2), t2 NO SE INSERTARÁ. 
El segundo INSERT dará un error que debes reportar en el archivo de violaciones a las RI.

**Comment by anticarasosman** [comment]
Hola disculpa ¿Hay un largo maximo o minimo para el numero de alumno?

**Reply by ypne** [reply]
No hay, solo es un entero

**Comment by AmarantaB** [comment]
Estimado, quisiera consultarle si podría aclararme el significado de la jerarquización en la hoja Planeación. Encontramos valores del tipo sigla, no obstante no sabemos a que hace referencia y como utilizarlo, desde ya muchas gracias. 

![image](https://github.com/user-attachments/assets/59220022-3a59-4c71-bac9-0cc7ec1df45b)
